ref/haskell/src/Lib.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
```

Comment:
I could, but Data.IntMap and Data.Map are in a different dependency `containers`. Same for vectors and arrays. I could use lists (standard Haskell lists are linked lists :cry:), but access would be O(n) then. Both `containers` and `unordered-containers` (which I use here) are part of Haskell Platform though, which means it is commonly used (or at least deemed to be).